src/shared/mount-util.c

```diff
@@ -1698,17 +1700,17 @@ int bind_mount_submounts(
         return ret;
 }
 
-int make_mount_point_inode_from_stat(const struct stat *st, const char *dest, mode_t mode) {
-        assert(st);
+int make_mount_point_inode_from_mode(int dir_fd, const char *dest, mode_t source_mode, mode_t target_mode) {
+        assert(dir_fd >= 0 || dir_fd == AT_FDCWD);
         assert(dest);
 
-        if (S_ISDIR(st->st_mode))
-                return mkdir_label(dest, mode);
+        if (S_ISDIR(source_mode))
+                return mkdirat_label(dir_fd, dest, target_mode);
         else
-                return RET_NERRNO(mknod(dest, S_IFREG|(mode & ~0111), 0));
+                return RET_NERRNO(mknodat(dir_fd, dest, S_IFREG|(target_mode & ~0111), 0));
```

Comment:
Mixing `S_IFREG` into some file types can create invalid file types. For example, mixing `S_IFREG 0100000 ` into `S_FIFO 0010000` will create `mode_t` with file type which is not defined. 

Btw, this bug existed in the code also before.

I think the 3rd argument to `mknodat()` should be something like, `S_IFREG|(target_mode & ~0770111)`

Reply:
well, the assumption here is that the parameter is not passed with an ftype set, i.e. only the lower 12 bit are possibly set. and in all our uses that indeed happens.

but i guess we can mask this of just for paranoia